File: frontend/lib/shared/models/request_activity_event.dart
import 'package:flutter/material.dart';

import '../theme/app_colors.dart';
import '../widgets/app_icons.dart';

/// One line in the life of a request the care team shares.
///
/// The patient sees this as well as staff, which is the whole reason it exists
/// separately from the audit log: a status chip can say "pending" for three
/// days without ever answering "is anyone actually looking at this?".
enum RequestActivityAction {
  opened,
  claimed,
  released,
  escalated,
  resolved,
  declined,
  cancelled,
  note,
}

extension RequestActivityActionX on RequestActivityAction {
  IconData get icon => switch (this) {
    RequestActivityAction.opened => AppIcons.send,
    RequestActivityAction.claimed => AppIcons.acknowledge,
    RequestActivityAction.released => AppIcons.refresh,
    RequestActivityAction.escalated => AppIcons.trendUp,
    RequestActivityAction.resolved => AppIcons.check,
    RequestActivityAction.declined => AppIcons.error,
    RequestActivityAction.cancelled => AppIcons.close,
    RequestActivityAction.note => AppIcons.notes,
  };

  Color get color => switch (this) {
    RequestActivityAction.opened => AppColors.info,
    RequestActivityAction.claimed => AppColors.brandIndigo,
    RequestActivityAction.released => AppColors.warning,
    RequestActivityAction.escalated => AppColors.mcareAmber,
    RequestActivityAction.resolved => AppColors.success,
    RequestActivityAction.declined => AppColors.critical,
    RequestActivityAction.cancelled => AppColors.textMutedAA,
    RequestActivityAction.note => AppColors.textMutedAA,
  };

  /// Reads as `who verb`, so the sentence is built at the call site with the
  /// actor's name in front.
  String get verb => switch (this) {
    RequestActivityAction.opened => 'raised this request',
    RequestActivityAction.claimed => 'took it on',
    RequestActivityAction.released => 'handed it back to the team',
    RequestActivityAction.escalated => 'escalated it to care admin',
    RequestActivityAction.resolved => 'completed it',
    RequestActivityAction.declined => 'could not provide it',
    RequestActivityAction.cancelled => 'withdrew the request',
    RequestActivityAction.note => 'added a note',
  };
}

class RequestActivityEvent {
  const RequestActivityEvent({
    required this.id,
    required this.action,
    required this.actorLabel,
    required this.happenedAt,
    this.note,
  });

  final String id;
  final RequestActivityAction action;

  /// Frozen at write time on the server, so the trail still reads correctly
  /// after a clinician leaves.
  final String actorLabel;
  final DateTime happenedAt;
  final String? note;

  String get sentence => '$actorLabel ${action.verb}';

  static RequestActivityEvent fromApi(Map<String, dynamic> json) =>
      RequestActivityEvent(
        id: (json['id'] ?? '').toString(),
        action: RequestActivityAction.values.firstWhere(
          (a) => a.name == json['action'],
          orElse: () => RequestActivityAction.note,
        ),
        actorLabel: (json['actor_label'] ?? 'Someone') as String,
        happenedAt:
            DateTime.tryParse(
              (json['happened_at'] ?? '') as String,
            )?.toLocal() ??
            DateTime.now(),
        note: json['note'] as String?,
      );

  static List<RequestActivityEvent> listFromApi(Object? raw) =>
      (raw as List? ?? const [])
          .map((e) => fromApi((e as Map).cast<String, dynamic>()))
          .toList(growable: false);
}

File: backend/app/Models/RequestActivityEvent.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\MorphTo;

/**
 * One line in the life of a request the care team shares.
 *
 * Kept separate from `audit_entries` on purpose: the audit log records what
 * staff did for staff and compliance, while this trail is shown to the patient
 * as well — it is the answer to "is anyone actually looking at this?", which a
 * status field alone could never give.
 *
 * The actor's name is frozen into `actor_label` at write time so the timeline
 * still reads correctly after a clinician leaves the practice.
 */
class RequestActivityEvent extends Model
{
    public const OPENED = 'opened';
    public const CLAIMED = 'claimed';
    public const RELEASED = 'released';
    public const ESCALATED = 'escalated';
    public const RESOLVED = 'resolved';
    public const DECLINED = 'declined';
    public const CANCELLED = 'cancelled';
    public const NOTE = 'note';

    protected $fillable = [
        'subject_type',
        'subject_id',
        'actor_user_id',
        'actor_label',
        'action',
        'note',
        'happened_at',
    ];

    protected function casts(): array
    {
        return ['happened_at' => 'datetime'];
    }

    public function subject(): MorphTo
    {
        return $this->morphTo();
    }

    /**
     * Appends an event to [$subject]'s trail.
     *
     * Never throws. A missing timeline line is worth less than the action it
     * describes, and a claim that 500s after the claim has been written leaves
     * two clinicians unsure which of them holds the request.
     */
    public static function record(
        Model $subject,
        string $action,
        string $actorLabel,
        ?int $actorUserId = null,
        ?string $note = null,
    ): void {
        try {
            static::create([
                'subject_type' => $subject->getMorphClass(),
                'subject_id' => $subject->getKey(),
                'actor_user_id' => $actorUserId,
                'actor_label' => $actorLabel,
                'action' => $action,
                'note' => $note,
                'happened_at' => now(),
            ]);
        } catch (\Throwable $e) {
            report($e);
        }
    }

    public function toApiArray(): array
    {
        return [
            'id' => (string) $this->id,
            'action' => $this->action,
            'actor_label' => $this->actor_label,
            'note' => $this->note,
            'happened_at' => $this->happened_at?->toIso8601String(),
        ];
    }
}
